posix: eventfd: fix read deadlock and add test coverage

cfriedt opened the pull request:
7 commits (please see each commit message for details)

* tests: posix: eventfd: use reopen in test_write_then_read (should have gone into #58406)
* lib: os: add `k_condvar` for each fdtable entry
* posix: eventfd: un-inline `eventfd_read()` and `eventfd_write()`
* posix: eventfd: revise locking, signaling, and allocation
* tests: posix: eventfd: blocking read-write test for deadlock bug
* posix: eventfd: deprecate non-public EFD macros
* tests: posix: eventfd: add stress test

Fixes #58426

Note: this improves non-blocking `eventfd` performance significantly.

I would suggest reading each commit individually with the exception of the eventfd revision.

For that one, it's probably easiest just to look at the end result [here](https://github.com/zephyrproject-rtos/zephyr/blob/5d67c8a8037d0c6b974cdc45ae2e779fae8a5c42/lib/posix/eventfd.c).

Commit:
tests: eventfd: blocking read-write test for deadlock bug

Previously, the prototypical use case for `eventfd()` was not
possible in Zephyr due to a deadlock associated with the
`fdtable.c` `k_mutex`.

The prototypical use case for `eventfd()` is to make a
blocking call to `eventfd_read()` from one thread, and then
from another thread, call `eventfd_write()` to "wake up" the
reading thread.

This additional test case ensures that Zephyr supports the
prototypical `eventfd()` use case.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: eventfd: un-inline eventfd_read() and eventfd_write()

There is no major benefit to haveing `eventfd_read()` and
`eventfd_write()` as `static inline` functions and it is
arguably much easier to solve the deadlock bug when they
are not `inline`.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
lib: os: add k_condvar for each fdtable entry

Complement the `struct k_mutex` in each fdtable entry
with a `struct k_condvar`. The reasoning for this should be
self-evident.

For a bit of history, `fdtable` was introduced in
commit 06eb489c4573 ("kernel: add condition variables")
which predates `struct k_condvar`, introduced in
commit f484bbaa264f ("lib: posix: Implement generic file descriptor table")
by almost 2 years.

Additionally, provide a new accessor function,
`z_get_obj_lock_and_cond()`, that (optionally) gets the mutex
and condition variable associated with the provided object and
vtable.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: eventfd: revise locking, signaling, and allocation

TL;DR - a complete rewrite.

Previously, the prototypical `eventfd()` usage (one thread
performing a blocking `read()`, followed by another thread
performing a `write()`) would deadlock Zephyr. This shortcoming
has existed in Zephyr's `eventfd()` implementation from the
start and the suggested workaround was to use `poll()`.

However, that is not sufficient for integrating 3rd-party
libraries that may rely on proper `eventfd()` blocking
operations such as `eventfd_read()` and `eventfd_write()`.

The culprit was the per-fdtable-entry `struct k_mutex`.

Here we perform a minor revision of the locking strategy
and employ `k_condvar_broadcast()` and `k_condvar_wait()`
to signal and wait on the holder of a given `struct k_mutex`.

It is important to note, however, that the primary means of
synchronizing the eventfd state is actually the eventfd
spinlock. The fdtable mutex and condition variable are mainly
used for the purposes of blocking io (r,w,close) and are not
used in the code path of non-blocking reads.

The `wait_q` and `k_poll_signal` entries were removed from
`struct eventfd` as they were unnecessary.

Additionally, switch to using a bitarray because it is
possibly faster than linear search for allocating and
deallocating eventfd resources.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented on lib/posix/eventfd.c:
```diff
@@ -239,6 +134,170 @@ static int eventfd_ioctl_op(void *obj, unsigned int request, va_list args)
+	return 0;
+}
+
+static ssize_t eventfd_write_op(void *obj, const void *buf, size_t sz)
```
Needs to return the number of bytes written on success.

cfriedt commented on lib/posix/eventfd.c:
```diff
@@ -239,6 +134,170 @@ static int eventfd_ioctl_op(void *obj, unsigned int request, va_list args)
+	return 0;
+}
+
+static ssize_t eventfd_read_op(void *obj, void *buf, size_t sz)
```
Needs to return the number of bytes read on success, not 0

cfriedt commented on lib/posix/eventfd.c:
```diff
@@ -248,49 +307,56 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
-	k_poll_signal_raise(&efd->write_sig, 0);
 
-	z_finalize_fd(fd, efd, &eventfd_fd_vtable);
+	return eventfd_read_op(obj, value, sizeof(*value));
```
Return 0 on success (i.e. if `eventfd_read_op()` reads `sizeof(*value)` bytes)

Commit:
tests: posix: eventfd: use reopen in test_write_then_read

Use `reopen()` in one additional place in the eventfd
testsuite.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* added Meta copyright since `lib/posix/eventfd.c` was more or less a complete rewrite

cfriedt commented:
This should probably be merged after #58743

cfriedt commented:
Marking `DNM` for today.

I'm going to do another respin of this with special consideration given to the case where `O_NONBLOCK` is specified, but also taking into account some of the potential race conditions that could occur between threads attempting to access the same file descriptor - e.g. `poll()` vs `close()`.

andyross reviewed (dismissed):
Looks good.  I didn't do a hugely deep dive into the synchronization strategy, but the description makes sense and the code seems to match.

Commit:
posix: eventfd: deprecate non-public EFD macros

Deprecate `EFD_IN_USE` and `EFD_FLAGS_SET` as they are not
specified as part of any public `eventfd()` API.

While those are being deprecated, use `_INTERNAL` variants.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
I think that will do - PTAL @andyross, when you have a minute.

```shell
twister -i -p mps2_an385 -p qemu_riscv64 -p qemu_cortex_a53 -p qemu_x86_64 -p qemu_riscv64_smp -p qemu_cortex_a53_smp -T tests/posix/eventfd/
Renaming output directory to /home/cfriedt/zephyrproject/zephyr/twister-out.5
INFO    - Using Ninja..
INFO    - Zephyr version: v3.4.0-rc1-101-g5d67c8a8037d
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    6/   6  100%  skipped:    0, failed:    0, error:    0
INFO    - 1 test scenarios (6 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO    - 6 of 6 test configurations passed (100.00%), 0 failed, 0 errored, 0 skipped with 0 warnings in 9.52 seconds
INFO    - In total 60 test cases were executed, 0 skipped on 6 out of total 573 platforms (1.05%)
INFO    - 6 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

cfriedt commented:
I would suggest reading each commit individually with the exception of the eventfd revision.

For that one, it's probably easiest just to look at the end result [here](https://github.com/zephyrproject-rtos/zephyr/blob/5d67c8a8037d0c6b974cdc45ae2e779fae8a5c42/lib/posix/eventfd.c).

jgl-meta commented on lib/posix/eventfd.c:
```diff
@@ -246,51 +262,181 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
+		/* retake the spinlock */
+		key = k_spin_lock(&efd->lock);
+
+		ret = op(efd, buf);
```
How safe is it to allow a caller defined function to be called while holding the spin lock?

I guess it is somewhat moot since it is clearly necessary to lock here. Just asking for curiosity.

cfriedt replied:
This function is static and the only functions that are being passed in are `eventfd_read_locked()` and `eventfd_wrote_locked()`

jgl-meta commented on lib/posix/eventfd.c:
```diff
@@ -246,51 +262,181 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
+
+	/* do not hold a spinlock when taking a mutex */
+	k_spin_unlock(&efd->lock, key);
+	err = k_mutex_lock(lock, K_FOREVER);
```
I think these next few lines of code could lead to one of the classic examples for deadlock 

t1: unlock(a) lock(b) lock(a)

and if another thread tries to acquire both locks then:
t2: lock(a) lock(b)
= deadlock

Is this scenario possible here? 

That said I think it is clear that you have to give up the spin lock before trying to lock the mutex. Just curious if the scenario is possible.

cfriedt replied:
> Is this scenario possible here? 

I can see why you might think that based on your earlier question.

Since `op()` is only either `eventfd_read_locked()` or 
`eventfd_write_locked()` any threads executing those functions will never be stalled on anything, so you can ignore the inner spin lock / unlock.


What you're left with is just the outer muted lock / unlock, which is business as usual.

jgl-meta commented on lib/posix/eventfd.c:
```diff
@@ -246,51 +262,181 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
+			goto unlock_mutex;
 		}
+
+		/* do not hold a spinlock when taking a mutex */
```
nit: while waiting of a condition variable maybe would be better?

cfriedt replied:
You take the muted as part of waiting on the condition variable

jgl-meta commented on lib/posix/eventfd.c:
```diff
@@ -246,51 +262,181 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
+
+	ret = eventfd_rw_op(obj, &value, sizeof(eventfd_t), eventfd_write_locked);
+	__ASSERT_NO_MSG(ret == -1 || ret == sizeof(eventfd_t));
+	if (ret < 0) {
```
Is this superfluous? Does it have to be -1 can you just return ret?

cfriedt replied:
The return value needs to be either 0 or -1, but `ret` will be either -1 or `sizeof(eventfd_t)`.

Technically, I could have made a comparison against 0 as well, I just tend to prefer branching in the error case.

jgl-meta reviewed (approved):
A few small questions and comments. But I don't think anything to hold this back for if it addresses the problem.

jgl-meta reviewed (dismissed):
A few small questions and comments. But I don't think anything to hold this back for if it addresses the problem.

cfriedt commented:
> A few small questions and comments. But I don't think anything to hold this back for if it addresses the problem.

Hehe.. it's definitely functional. I would guess that it's also reasonably fast too, especially in the nong-blocking case.

Commit:
tests: posix: eventfd: add stress test

This test simply counts how many times `eventfd_read()`
and `eventfd_write()` can be called on an
`eventfd(0, EFD_SEMAPHORE | EFD_NONBLOCK)` file
descriptor.

Prior to the recent changes in `eventfd`, we were seeing
approximately < 1000 writes / s. However, the previous
`eventfd` implementation would fail this test with the
result that the number of successful reads was far greater
than the number of successful writes.

This should be impossible, and with the recent `eventfd`
changes that was fixed. Additionally, we are seeing an
increase in over 40x for non-blocking eventfd reads and
writes.

```
START - test_stress
I: BOARD: qemu_riscv64_smp
I: TEST_DURATION_S: 5
I: UPDATE_INTERVAL_S: 1
I: avg: 48537 reads/s
I: avg: 48575 writes/s
PASS - test_stress in 5.002 seconds
```

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
@andyross - you might be impressed by the performance improvement 😁 
@jgl-meta - PTAL
@keith-packard, @stephanosio - PTAL if you're interested as well

cfriedt commented:
Was rushing out the door and forgot to update the last commit. Stuck in horrible traffic right now 😭

cfriedt commented:
With `-DCONFIG_ASSERT=n` even higher throughput.
```shell
START - test_stress
I: BOARD: qemu_riscv64_smp
I: TEST_DURATION_S: 5
I: UPDATE_INTERVAL_S: 1
I: avg: 135537 writes/s
I: avg: 135472 reads/s
 PASS - test_stress in 5.002 seconds
```

I'll see if I can produce a flame graph before and after these changes... could be tricky though given that `CONFIG_ARCH_POSIX` and `CONFIG_POSIX_API` are still mutually exclusive.

cfriedt commented:
Remove logging from the testsuite, because it resulted in Compliance false positives.

cfriedt commented:
Probably the best apples-to-apples performance comparison can be made with this, before and after the changeset.
```shell
west build -p auto -b qemu_riscv64_smp -t run tests/posix/eventfd/ -- -DCONFIG_TEST_EXTRA_QUIET=y -DCONFIG_ASSERT=n
```

Before:
```shell
START - test_stress
BOARD: qemu_riscv64_smp
TEST_DURATION_S: 5
UPDATE_INTERVAL_S: 1
avg: 11093 reads/s
avg: 11094 writes/s
 PASS - test_stress in 4.998 seconds
```

After:
```shell
START - test_stress
BOARD: qemu_riscv64_smp
TEST_DURATION_S: 5
UPDATE_INTERVAL_S: 1
avg: 101623 reads/s
avg: 101624 writes/s
 PASS - test_stress in 4.999 seconds
```

```shell
echo "101623/11093" | bc -l
9.16100243396736680789
```

So just under a 10x performance improvement at the end of the day.

nashif commented on include/zephyr/sys/fdtable.h:
```diff
@@ -107,6 +107,24 @@ void *z_get_fd_obj(int fd, const struct fd_op_vtable *vtable, int err);
+ *
+ * @return `true` on success, `false` otherwise.
+ */
+bool z_get_obj_lock_and_cond(void *obj, const struct fd_op_vtable *vtable, struct k_mutex **lock,
```
z_ is reserved for private functions and calls, this header is full of z_ call and it is a public API, we need to fix this at some point.

cfriedt replied:
cc @jgl-meta

cfriedt replied:
>  header is full of z_ call and it is a public API, we need to fix this at some point. 

@nashif - what do you think of "zvfs" or "zfdio"?

andyross reviewed (commented):
Missed the +1 before it merged, but just for the record I can't see anything wrong with the synchronization.

cfriedt commented on tests/posix/eventfd/Kconfig:
```diff
@@ -0,0 +1,43 @@
+	range 1 21600
+	default 5
+	help
+	   Duration for the test, in seconds. The range has a reblatively high
```
"reblatively" - never fails that there will be at least one typo 😅
